src/partition/ReceivedPartition.cpp

```diff
@@ -469,144 +470,345 @@ void ReceivedPartition::createOwnerInformation()
-      if (_mesh->vertices()[i].isTagged()) {
-        masterAtInterface = true;
-        slaveTags[0][i]   = 1;
-      } else {
-        slaveTags[0][i] = 0;
+      else {
+        tags[i] = 0;
       }
     }
-    PRECICE_DEBUG("My tags: {}", slaveTags[0]);
 
-    // receive slave data
-    int ranksAtInterface = 0;
-    if (masterAtInterface)
-      ranksAtInterface++;
+    // #4: Exchange number of already owned vertices with the neighbors
 
-    for (int rank : utils::MasterSlave::allSlaves()) {
-      int localNumberOfVertices = -1;
-      utils::MasterSlave::_communication->receive(localNumberOfVertices, rank);
-      PRECICE_DEBUG("Rank {} has {} vertices.", rank, localNumberOfVertices);
-      slaveOwnerVecs[rank].resize(localNumberOfVertices, 0);
-
-      if (localNumberOfVertices != 0) {
-        PRECICE_DEBUG("Receive tags from slave rank {}", rank);
-        utils::MasterSlave::_communication->receive(slaveTags[rank], rank);
-        utils::MasterSlave::_communication->receive(slaveGlobalIDs[rank], rank);
-        PRECICE_DEBUG("Rank {} has tags {}", rank, slaveTags[rank]);
-        PRECICE_DEBUG("Rank {} has global IDs {}", rank, slaveGlobalIDs[rank]);
-        bool atInterface = false;
-        utils::MasterSlave::_communication->receive(atInterface, rank);
-        if (atInterface)
-          ranksAtInterface++;
+    // to store receive requests.
+    std::vector<com::PtrRequest> vertexNumberRequests;
+
+    // Define and initialize for load balancing
+    std::map<int, int> neighborRanksVertexCount;
+    for (auto &neighborRank : localConnectedBBMap) {
+      neighborRanksVertexCount.insert(std::make_pair(neighborRank.first, 0));
+    }
+
+    // Asynchronous recieve number of owned vertices from neighbor ranks
+    for (auto &neighborRank : localConnectedBBMap) {
+      auto request = utils::MasterSlave::_communication->aReceive(neighborRanksVertexCount.at(neighborRank.first), neighborRank.first);
+      vertexNumberRequests.push_back(request);
+    }
+
+    // Synchronous send number of owned vertices to neighbor ranks
+    for (auto &neighborRank : localConnectedBBMap) {
+      utils::MasterSlave::_communication->send(counter, neighborRank.first);
+    }
+
+    // wait until aReceive is complete.
+    for (auto &rqst : vertexNumberRequests) {
+      rqst->wait();
+    }
+
+    // #5: Exchange list of shared vertices with the neighbor
+
+    // to store send requests.
+    std::vector<com::PtrRequest> vertexListRequests;
+
+    for (auto &receivingRank : sharedVerticesSendMap) {
+      int  sendSize = receivingRank.second.size();
+      auto request  = utils::MasterSlave::_communication->aSend(sendSize, receivingRank.first);
+      vertexListRequests.push_back(request);
+      if (sendSize != 0) {
+        auto request = utils::MasterSlave::_communication->aSend(receivingRank.second, receivingRank.first);
+        vertexListRequests.push_back(request);
       }
```

Comment:
This logic can be implemented in `com::Communication` (the base class).
We could use an additional `out` iterator to save the requests:

Something along the lines of
```cpp
template<typename OutIter>
void Communication::aSend(const std::vector<int>& v , int rank, OutIter out) {
  //  ...
  *(out++) = sSend(v.size(), receivingRank.first);
  // ...
  *(out++) = aSend(v.data(), v.size(), receivingRank.first);
  // ---
}


utils::MasterSlave::_communication->aSend(receivingRank.second, receivingRank.first, std::back_inserter(vertexListRequests));
```

Reply:
I agree in general, but let's do it after pull request.